File: retab.rb
#!/usr/bin/env ruby
#
# author: scott w olesen <swo@mit.edu>

require 'arginine'
require 'fileutils'
require 'tempfile'

params = Arginine::parse do |a|
  desc "replace whitespace"
  opt :joiner, :short => "F", :default => "\t"
  opt :in_place, :desc => "edit following file in place"
  argf
end

if params[:in_place]
  tf = Tempfile.new("retab")
  open(params[:in_place]).each { |line| tf.puts line.split.join(params[:joiner]) }
  tf.close
  FileUtils.mv(tf.path, params[:in_place])
else
  # process from argf
  ARGF.each { |line| puts line.split.join(params[:joiner]) }
end

File: retab.py
#!/usr/bin/env python
#
# author: scott olesen <swo@mit.edu>

import argparse, re, fileinput

if __name__ == '__main__':
    p = argparse.ArgumentParser(description="replace whitespace with tabs", formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    p.add_argument('--regex', '-r', metavar='PATTERN', default='\s+', help="regex used to split line")
    p.add_argument('--in_place', '-i', metavar='EXT', nargs='?', default=False, help="edit file in place? use '.bak' if no EXT specified")
    p.add_argument('--delimiter', '-F', metavar='SEP', default="\t", help="output separator (the famous tab)")
    p.add_argument('input', nargs='*', metavar='FILE')
    args = p.parse_args()

    # in the case of "retab.py -i file", argparse thinks that "file" is the option
    # for -i, when it's actually the value for input, since when using -i, you must
    # take at least one input file
    if args.input == [] and args.in_place != False:
        args.input.append(args.in_place)
        args.in_place = None

    if args.in_place == False:
        kwargs = {}
    else:
        if args.in_place is None:
            kwargs = {'inplace': True, 'backup': '.bak'}
        else:
            kwargs = {'inplace': True, 'backup': '.' + args.in_place}

    for line in fileinput.input(args.input, **kwargs):
        fields = re.split(args.regex, line)
        new_line = args.delimiter.join(fields)
        print(new_line)
